wenet/transformer/asr_model.py

```diff
@@ -483,11 +515,36 @@ def attention_rescoring(
                                   encoder_out.size(1),
                                   dtype=torch.bool,
                                   device=device)
-        decoder_out, _ = self.decoder(
-            encoder_out, encoder_mask, hyps_pad,
-            hyps_lens)  # (beam_size, max_hyps_len, vocab_size)
+        if hasattr(self.decoder, 'left_decoder'):
+            # bitransformer decoder, the left to right decoder
+            decoder_out, _, _ = self.decoder.left_decoder(
+                encoder_out, encoder_mask, hyps_pad,
+                hyps_lens)  # (beam_size, max_hyps_len, vocab_size)
+        else:
+            # transformer decoder
+            decoder_out, _, _ = self.decoder(
+                encoder_out, encoder_mask, hyps_pad,
+                hyps_lens)  # (beam_size, max_hyps_len, vocab_size)
         decoder_out = torch.nn.functional.log_softmax(decoder_out, dim=-1)
         decoder_out = decoder_out.cpu().numpy()
+
+        # used for right to left decoder
+        if reverse_weight > 0:
+            assert hasattr(self.decoder, 'right_decoder')
+            r_hyps_pad = pad_sequence([(torch.flip(y.int()[:i], [0]))
+                                       for y, i in zip(ori_hyps_pad, hyps_lens)],
+                                      True, -1.0)
+            r_hyps_pad, _ = add_sos_eos(r_hyps_pad, self.sos, self.eos,
+                                        self.ignore_id)
+            r_decoder_out, _, _ = self.decoder.right_decoder(
+                encoder_out,
+                encoder_mask,
+                r_hyps_pad,
+                hyps_lens,
+            )  # (beam_size, max_hyps_len, vocab_size)
```

Comment:
We should think about how to abstract this into a unified representation, otherwise every time we do a forward, we have to use branch for the two cases.